File: puzzle6/python/puzzle6.py
"""Day 6 Puzzle"""

import re

class LightGrid(object):
    """Represents a grid of lights."""

    def __init__(self, width, height):
        """Constructor."""
        self._grid = []
        self._width = width
        self._height = height
        self.clear_grid()

    def clear_grid(self):
        """Make grid all zeros."""
        self._grid = []
        for dummy_index in range(self._height):
            self._grid.append([0 for dummy_kndex in range(self._width)])

    def __str__(self):
        """__str__ method."""
        result = ""
        for row in self._grid:
            result = result + str(row) + "\n"
        return result

    def toggle_light(self, row, col):
        """Toggles lights on and off."""
        if self._grid[row][col] == 1:
            self._grid[row][col] = 0
        else:
            self._grid[row][col] = 1

    def toggle_rectangle_of_lights(self, row1, col1, row2, col2):
        """Toggles lights from [row1, col1] to [row2, col2] inclusive."""
        for row in range(row1, row2 + 1):
            for col in range(col1, col2 + 1):
                self.toggle_light(row, col)

    def turn_on_light(self, row, col):
        """Turns on light."""
        self._grid[row][col] = 1

    def turn_on_rectangle_of_lights(self, row1, col1, row2, col2):
        """Turns on lights from [row1, col1] to [row2, col2] inclusive."""
        for row in range(row1, row2 + 1):
            for col in range(col1, col2 + 1):
                self.turn_on_light(row, col)

    def turn_off_light(self, row, col):
        """Turns off light."""
        self._grid[row][col] = 0

    def turn_off_rectangle_of_lights(self, row1, col1, row2, col2):
        """Turns off lights from [row1, col1] to [row2, col2] inclusive."""
        for row in range(row1, row2 + 1):
            for col in range(col1, col2 + 1):
                self.turn_off_light(row, col)

    def get_grid(self):
        """Returns the list of lists that is the grid."""
        return self._grid

    def count_lights_on(self):
        """Returns the number of lights on."""
        return sum([sum(row) for row in self._grid])

    def process_instruction(self, line):
        """Process a text instruction."""
        tokens = [item for item in re.split(r'(\W+)', line) \
                  if item not in [' ', ',']]
        if tokens[0] == 'turn':
            if tokens[1] == 'on':
                self.turn_on_rectangle_of_lights(int(tokens[2]), int(tokens[3]),
                                                 int(tokens[5]), int(tokens[6]))
            elif tokens[1] == 'off':
                self.turn_off_rectangle_of_lights(int(tokens[2]), int(tokens[3]),
                                                  int(tokens[5]), int(tokens[6]))
        elif tokens[0] == 'toggle':
            self.toggle_rectangle_of_lights(int(tokens[1]), int(tokens[2]),
                                            int(tokens[4]), int(tokens[5]))



class AdjustableLightGrid(LightGrid):
    """Represents grid of lights with adjustable brightness."""

    def __init__(self, width, height):
        """Constructor."""
        LightGrid.__init__(self, width, height)

    def toggle_light(self, row, col):
        self._grid[row][col] += 2

    def turn_off_light(self, row, col):
        if self._grid[row][col] > 0:
            self._grid[row][col] -= 1

    def turn_on_light(self, row, col):
        self._grid[row][col] += 1


def process_input_file(file_name, grid):
    """Process an input file."""
    input_file = open(file_name, 'r')
    for line in input_file:
        grid.process_instruction(line.strip())
    input_file.close()
    return grid.count_lights_on()

if __name__ == '__main__':
    print 'There are', \
          process_input_file('../input.txt', LightGrid(1000, 1000)), \
          'lights on'
    print 'The total brightness is', \
          process_input_file('../input.txt', AdjustableLightGrid(1000, 1000))

File: puzzle6/go/puzzle6.go
// Advent of Code 2015 :: Day 6: Probably a Fire Hazard

package main

import (
    "bufio"
    "fmt"
    "os"
    "strconv"
    "strings"
)

type Command struct {
    cmd string
    row1 int
    col1 int
    row2 int
    col2 int
}

func turnOn(row1 int, col1 int, row2 int, col2 int, grid [][]bool) {
    for r := row1; r <= row2; r += 1 {
        for c := col1; c <= col2; c += 1 {
            grid[r][c] = true
        }
    }
}

func toggle(row1 int, col1 int, row2 int, col2 int, grid [][]bool) {
    for r := row1; r <= row2; r += 1 {
        for c := col1; c <= col2; c += 1 {
            grid[r][c] = !grid[r][c]
        }
    }
}

func turnOff(row1 int, col1 int, row2 int, col2 int, grid [][]bool) {
    for r := row1; r <= row2; r += 1 {
        for c := col1; c <= col2; c += 1 {
            grid[r][c] = false
        }
    }
}

func countLightsOn(grid [][]bool) int {
    lightsOn := 0
    for r := 0; r < 1000; r += 1 {
        for c := 0; c < 1000; c += 1 {
            if grid[r][c] {
                lightsOn += 1
            }
        }
    }
    return lightsOn
}

func turnUp(row1 int, col1 int, row2 int, col2 int, grid [][]int) {
    for r := row1; r <= row2; r += 1 {
        for c := col1; c <= col2; c += 1 {
            grid[r][c] += 1
        }
    }
}

func turnWayUp(row1 int, col1 int, row2 int, col2 int, grid [][]int) {
    for r := row1; r <= row2; r += 1 {
        for c := col1; c <= col2; c += 1 {
            grid[r][c] += 2
        }
    }
}

func turnDown(row1 int, col1 int, row2 int, col2 int, grid [][]int) {
    for r := row1; r <= row2; r += 1 {
        for c := col1; c <= col2; c += 1 {
            grid[r][c] -= 1
            if (grid[r][c] < 0) {
                grid[r][c] = 0
            }
        }
    }
}

func countBrightness(grid [][]int) int {
    brightness := 0
    for r := 0; r < 1000; r += 1 {
        for c := 0; c < 1000; c += 1 {
            brightness += grid[r][c]
        }
    }
    return brightness
}

func solve1(commands []Command) int {
    // Create grid
    grid := make([][]bool, 1000)
    for i := 0; i < 1000; i += 1 {
        grid[i] = make([]bool, 1000)
    }

    for _, command := range commands {
        switch command.cmd {
        case "turn off":
            turnOff(command.row1, command.col1, command.row2, command.col2, grid)
        case "turn on":
            turnOn(command.row1, command.col1, command.row2, command.col2, grid)
        case "toggle":
            toggle(command.row1, command.col1, command.row2, command.col2, grid)
        }
    }
    return countLightsOn(grid)
}

func solve2(commands []Command) int {
    // Create grid
    grid := make([][]int, 1000)
    for i := 0; i < 1000; i += 1 {
        grid[i] = make([]int, 1000)
    }

    for _, command := range commands {
        switch command.cmd {
        case "turn off":
            turnDown(command.row1, command.col1, command.row2, command.col2, grid)
        case "turn on":
            turnUp(command.row1, command.col1, command.row2, command.col2, grid)
        case "toggle":
            turnWayUp(command.row1, command.col1, command.row2, command.col2, grid)
        }
    }
    return countBrightness(grid)
}

func main() {

    commands := []Command{}

    scanner := bufio.NewScanner(os.Stdin)
    for scanner.Scan() {
        // Read
        line := scanner.Text()
        tokens := strings.Split(line, " ")
        // Parse
        cmd := tokens[0]
        if cmd == "turn" {
            cmd = cmd + " " + tokens[1]
        }
        start := 2
        end := 4
        if cmd == "toggle" {
            start = 1
            end = 3
        }
        posn := strings.Split(tokens[start], ",")
        row1, _ := strconv.Atoi(posn[0])
        col1, _ := strconv.Atoi(posn[1])
        posn = strings.Split(tokens[end], ",")
        row2, _ := strconv.Atoi(posn[0])
        col2, _ := strconv.Atoi(posn[1])

        command := Command{cmd, row1, col1, row2, col2}
        commands = append(commands, command)
    }
    fmt.Println("The solution to part 1 is", solve1(commands))
    fmt.Println("The solution to part 1 is", solve2(commands))
}
